File: 1.300/242.ValidAnagram.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <algorithm>

using namespace std;

//242. 有效的字母异位词
class ValidAnagram {
    /* 给定两个字符串 s 和 t ，编写一个函数来判断 t 是否是 s 的字母异位词。
    */
public:
    bool isAnagram(string s, string t) {
        if (s.size() != t.size())
            return false;
        //哈希比对
        unordered_map<char, int> src;
        unordered_map<char, int> dst;
        for (char c : s) {
            src[c] += 1;
        }
        for (char c : t) {
            dst[c] += 1;
        }
        for (auto& it : src) {
            if (it.second != dst[it.first])
                return false;
        }
        return true;
    }
public:
    bool isAnagram2(string s, string t) {
        //排序对比
        if (s.size() != t.size())
            return false;
        sort(s.begin(), s.end());
        sort(t.begin(), t.end());
        for (int i = 0;i < s.size(); ++i) {
            if (s[i] != t[i])
                return false;
        }
        return true;
    }
};

int main(int argc, char const* argv[]) {
    ValidAnagram cls;
    cout << boolalpha << cls.isAnagram("anagram", "nagaram") << "->true" << endl;
    cout << boolalpha << cls.isAnagram("rat", "car") << "->false" << endl;

    cout << boolalpha << cls.isAnagram2("anagram", "nagaram") << "->true" << endl;
    cout << boolalpha << cls.isAnagram2("rat", "car") << "->false" << endl;
    return 0;
}

File: go/1.300/242.ValidAnagram.go
package main

import "sort"

func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	src := map[rune]int{}
	dst := map[rune]int{}
	for _, v := range s {
		src[v]++
	}
	for _, v := range t {
		dst[v]++
	}
	for k, v := range src {
		if v1, ok := dst[k]; !ok || v != v1 {
			return false
		}
	}
	return true
}

func isAnagram2(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	src := []byte(s)
	dsc := []byte(t)
	sort.Slice(src, func(i, j int) bool {
		return src[i] < src[j]
	})
	sort.Slice(dsc, func(i, j int) bool {
		return dsc[i] < dsc[j]
	})
	for i, v := range src {
		if dsc[i] != v {
			return false
		}
	}
	return true
}

func main() {

}
